File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter15/ClockRack/ClockRack/DistributedUniformGrid.cs
using System;
using Windows.Foundation;
using Windows.UI.Xaml.Controls;
using Petzold.ProgrammingWindows6.Chapter11;

namespace ClockRack
{
    public class DistributedUniformGrid : UniformGrid
    {
        protected override Size ArrangeOverride(Size finalSize)
        {
            int index = 0;
            double cellWidth = finalSize.Width / cols;
            double cellHeight = finalSize.Height / rows;
            int displayed = 0;

            if (this.Orientation == Orientation.Vertical)
            {
                for (int row = 0; row < rows; row++)
                {
                    double y = row * cellHeight;
                    int accumDisplay = (int)Math.Ceiling((row + 1.0) * this.Children.Count / rows);
                    int display = accumDisplay - displayed;
                    cellWidth = Math.Round(finalSize.Width / display);
                    double x = 0;

                    for (int col = 0; col < display; col++)
                    {
                        if (index < this.Children.Count)
                            this.Children[index].Arrange(new Rect(x, y, cellWidth, cellHeight));

                        x += cellWidth;
                        index++;
                    }
                    displayed += display;
                }
            }
            else
            {
                for (int col = 0; col < cols; col++)
                {
                    double x = col * cellWidth;
                    int accumDisplay = (int)Math.Ceiling((col + 1.0) * this.Children.Count / cols);
                    int display = accumDisplay - displayed;
                    cellHeight = Math.Round(finalSize.Height / display);
                    double y = 0;

                    for (int row = 0; row < display; row++)
                    {
                        if (index < this.Children.Count)
                            this.Children[index].Arrange(new Rect(x, y, cellWidth, cellHeight));

                        y += cellHeight;
                        index++;
                    }
                    displayed += display;
                }
            }
            return finalSize;
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter11/PetzoldProgrammingWindows6Chapter11/PetzoldProgrammingWindows6Chapter11/DistributedUniformGrid.cpp
#include "pch.h"
#include "DistributedUniformGrid.h"

using namespace PetzoldProgrammingWindows6Chapter11;

using namespace Windows::Foundation;

DistributedUniformGrid::DistributedUniformGrid()
{
}

Size DistributedUniformGrid::ArrangeOverride(Size finalSize)
{
    unsigned int index = 0;
    float cellWidth = finalSize.Width / cols;
    float cellHeight = finalSize.Height / rows;
    int displayed = 0;

    if (this->Orientation == Windows::UI::Xaml::Controls::Orientation::Vertical)
    {
        for (int row = 0; row < rows; row++)
        {
            float y = row * cellHeight;
            int accumDisplay = (int)ceil((row + 1.0f) * this->Children->Size / rows);
            int display = accumDisplay - displayed;
            cellWidth = (float)(int)(finalSize.Width / display + 0.5f);
            float x = 0;

            for (int col = 0; col < display; col++)
            {
                if (index < this->Children->Size)
                    this->Children->GetAt(index)->Arrange(Rect(x, y, cellWidth, cellHeight));

                x += cellWidth;
                index++;
            }
            displayed += display;
        }
    }
    else
    {
        for (int col = 0; col < cols; col++)
        {
            float x = col * cellWidth;
            int accumDisplay = (int)ceil((col + 1.0f) * this->Children->Size / cols);
            int display = accumDisplay - displayed;
            cellHeight = (float)(int)(finalSize.Height / display + 0.5f);
            float y = 0;

            for (int row = 0; row < display; row++)
            {
                if (index < this->Children->Size)
                    this->Children->GetAt(index)->Arrange(Rect(x, y, cellWidth, cellHeight));

                y += cellHeight;
                index++;
            }
            displayed += display;
        }
    }
    return finalSize;
}


